### PR TITLE
checkJs swingset-vat

### PR DESCRIPTION
**Pull request by turadg**
refs: https://github.com/Agoric/agoric-sdk/issues/4885 (progress)

## Description

Extracted from https://github.com/Agoric/agoric-sdk/pull/6537

Changes isolated by commit to make review easier.

Note that this omits `test` from tsconfig so that `checkJs` can be enabled. Only one test file has `ts-check` so it's not much loss. To add `test` back in with `checkJs` would require fixing the "Found 938 errors in 76 files" or adding `@ts-nocheck` in files that aren't deemed worthy to fix (e.g. 161 errors in `test-liveslots.js`).

### Testing Considerations

CI

### CODE REVIEW

**Comment by mhofman** [comment]
I have a stash of type improvements to SwingSet I had after manually enabling type checks on some files. Will bring them along here.

File: packages/SwingSet/src/liveslots/liveslots.js
Comment (mhofman) [inline packages/SwingSet/src/liveslots/liveslots.js:10]: I believe this is erroneous, and `HandledPromise` is currently expected to be available on the global, installed by `@endo/init`, which is evaluated as part of the supervisor bundle before liveslots.

`HandledPromise` should correctly be typed as a global, maybe the file where this is declared doesn't get transitively imported?

File: packages/SwingSet/src/kernel/vat-loader/manager-subprocess-node.js
Comment (mhofman) [inline packages/SwingSet/src/kernel/vat-loader/manager-subprocess-node.js:81]: Ugh I hate this `void` incompatibility with `any` rule, and the runtime changes it requires. I wish there was an easy way to inform type params of generic functions in JSDoc. I may prefer if we "cast" the `makePromiseKit` call to `PromiseKit<void>`

File: packages/SwingSet/src/lib-nodejs/waitUntilQuiescent.js
Comment (mhofman) [inline packages/SwingSet/src/lib-nodejs/waitUntilQuiescent.js]: Same, can we cast the `makePromiseKit` result instead?

File: packages/SwingSet/tools/vat.js
Comment (mhofman) [inline packages/SwingSet/tools/vat.js:35]: That seems like a legitimate error that shouldn't silenced

File: packages/SwingSet/src/kernel/state/deviceKeeper.js
Comment (mhofman) [inline packages/SwingSet/src/kernel/state/deviceKeeper.js]: Please type the `res` declaration instead:

```js
    /** @type {Array<[string, string, string]>} */
    const res = [];
```

File: packages/SwingSet/src/kernel/state/reachable.js
Comment (mhofman) [inline packages/SwingSet/src/kernel/state/reachable.js]: That seems to be a legitimate mistake uncovered by tsc

```suggestion
    assert.fail(`flag (${flag}) must be 'R' or '_'`);
  }
```

File: packages/SwingSet/src/kernel/slogger.js
Comment (mhofman) [inline packages/SwingSet/src/kernel/slogger.js:285]: Ugh the console/logger mistypings were intractable last time I attempted some fixes. Hopefully someday we can sort them out.

File: packages/SwingSet/src/lib/storageAPI.js
Comment (mhofman) [inline packages/SwingSet/src/lib/storageAPI.js:167]: That seems like a potential issue. @gibson042 ?

File: packages/SwingSet/src/lib-nodejs/worker-protocol.js
Comment (mhofman) [inline packages/SwingSet/src/lib-nodejs/worker-protocol.js]: Maybe we should be explicit and call `buf.toString()` or `String(buf)` instead of relying on the implicit conversion?

File: packages/SwingSet/src/liveslots/liveslots.js
Comment (mhofman) [inline packages/SwingSet/src/liveslots/liveslots.js:1499]: @FUDCo sounds like `scanForDeadObjects` is missing a `return` ?

File: packages/SwingSet/src/vats/comms/state.js
Comment (mhofman) [inline packages/SwingSet/src/vats/comms/state.js]: @warner do we actually ever use `CapData` with a `Slot` type that isn't `string`? What is the purpose of making it generic?

File: packages/SwingSet/test/util.js
Comment (turadg) [inline packages/SwingSet/test/util.js:48]: optionality conveyed by jsdoc so they type checker doesn't take this as a hint that the value is only ever `undefined`

File: packages/SwingSet/src/vats/vat-admin/vat-vat-admin.js
Comment (mhofman) [inline packages/SwingSet/src/vats/vat-admin/vat-vat-admin.js]: Please type `producePRR` to return a tuple instead of an array, e.g. with `/** @type {const} */` on the return value.

File: packages/SwingSet/src/vats/vat-admin/vat-vat-admin.js
Comment (mhofman) [inline packages/SwingSet/src/vats/vat-admin/vat-vat-admin.js]: Same as above

File: packages/SwingSet/test/util.js
Comment (mhofman) [inline packages/SwingSet/test/util.js]: Type the params instead to avoid `iface` inference to `string` ?

**Review by mhofman — CHANGES_REQUESTED** [review]
Thanks for getting this started, I had meant to get better typing into SwingSet.

I would like to see some of the `ts-expect-error` handled differently. And I believe type checks actually exposed a couple legitimate issues that are not tagged as FIXME.

**Comment by mhofman** [comment]
> I have a stash of type improvements to SwingSet I had after manually enabling type checks on some files. Will bring them along here.

Here is a crude rebase of that stash for future reference (not advocating for it to be taken as-is): https://github.com/Agoric/agoric-sdk/commit/94065955903e513302acb67eb25ec4f075c13c71

It may be worth including cosmic-swingset in this PR as its so interleaved with swingset for all the init flow.

File: packages/SwingSet/src/liveslots/liveslots.js
Comment (warner) [reply packages/SwingSet/src/liveslots/liveslots.js:10]: liveslots currently gets imported by e.g. `src/supervsors/subprocess-xsnap/supervisor-subprocess-xsnap.js`, when building the supervisor bundle.

In the future, we might make a separate liveslots-only bundle, using `liveslots.js` as the entrypoint file.

Dunno how that impacts the type management. I concur with Mathieu that HandledPromise is generally expected to be a global. I don't see any problems with importing it as well, assuming we don't somehow wind up with a second version (identity discontinuity), but yeah it's probably best if it's just on the global.

File: packages/SwingSet/tools/vat.js
Comment (warner) [reply packages/SwingSet/tools/vat.js:35]: BTW `vat.js` is probably the oldest file we have, and it's surely bitrotted badly by now. I haven't tried to run it in years.

File: packages/SwingSet/src/kernel/state/reachable.js
Comment (warner) [reply packages/SwingSet/src/kernel/state/reachable.js]: oops, yeah, I should have written `assert.fail`

File: packages/SwingSet/src/liveslots/liveslots.js
Comment (warner) [reply packages/SwingSet/src/liveslots/liveslots.js:1499]: Hm, this looks like both functions think they're responsible for the re-scan.. I bet we didn't think this through completely when we introduced `scanForDeadObjects()`.

The `doMore` test is meant to deal with the case where `gcAndFinalize` shakes loose some in-RAM `Representative`s,  `scanForDeadObjects` looks through `possiblyDeadSet` and finds some actual vrefs it can retire (`vrm.possibleVirtualObjectDeath()`), some of those vrefs were used as keys in a (virtual) WeakMap, so deleting the keys can drop references to more in-RAM Representatives. We need a new `gcAndFinalize` pass to shake those loose, so we repeat the loop within `scanForDeadObjects()` until `possibleVirtualObjectDeath()` didn't free anything (and our other checks didn't free anything either).

The outer function(`bringOutYourDead`) looks like it was meant to do the same "repeat until nothing changes" pattern, although with recursion rather than a `do/while` loop. It does a (redundant?) extra `gcAndFinalize` at the expected time. I think the only real contribution of the outer loop, these days, is an extra call to `vom.flushCache()`, but I'm pretty sure that can't shake loose more objects like `doMore` is tracking, so I think it's not important to call from there.

@FUDCo is the authority here, but I'm thinking there are two correct fixes:

* 1: remove the `gcAndFinalize` and the recursion from `bringOutYourDead` (makie it just do `vom.flushCache(); await gcTools.gcAndFinalize(); await scanForDeadObjects(); return`, and rely upon the loop that happens inside `scanForDeadObjects`
* 2: remove the `do/while` from `scanForDeadObjects`, and modify it to return `doMore`, relying upon the recursion in `bringOutYourDead` to shake everything loose
  * if we go this way, I think I'd prefer to have `bringOutYourDead` use a do/while loop, instead of the recursion, since we can't rely on tail-call optimization

The benefit of doing the loop inside `scanForDeadObjects` (1) is that we only emit a single trio of syscalls, no matter how many times we have to loop. I can't think of a benefit for (2), other than that (1) makes `bringOutYourDead` start to look kind of useless and suggests we might refactor something

File: packages/SwingSet/src/vats/comms/state.js
Comment (warner) [reply packages/SwingSet/src/vats/comms/state.js]: The kernel only ever uses strings, although in some places they're krefs and in others they're vrefs, (and comms also has rrefs and lrefs). I bet it'd catch some potentials bugs if we had distinct types for those.

I think `CapData` is actually defined by `marshal`, which doesn't enforce `typeof(slot)==='string'`, so `CapData` is a generic (parameterized by some type T, which the kernel always fills with `string`). I think we have a more-specialized `KernelCapData` or `SwingsetCapData` within `packages/SwingSet/`, since we can't use the generic one.

I can't think of a current use case for non-string slots, so I'd probably be ok if `marshal` went ahead and assumed they were always strings. But I also expect that the `marshal` folks (@erights) would think of that as unnecessarily narrow / overly constrained.

**Comment by warner** [comment]
I'll defer to @mhofman for anything beyond the bits I mentioned above.

File: packages/SwingSet/src/liveslots/liveslots.js
Comment (turadg) [reply packages/SwingSet/src/liveslots/liveslots.js:10]: > should correctly be typed as a global

Granted. Given that this solves the PR scope (enable `checkJs`) and that importing it doesn't create any problems, I propose deferring that as future work.

File: packages/SwingSet/src/kernel/vat-loader/manager-subprocess-node.js
Comment (turadg) [reply packages/SwingSet/src/kernel/vat-loader/manager-subprocess-node.js:81]: Great idea. done

File: packages/SwingSet/tools/vat.js
Comment (turadg) [reply packages/SwingSet/tools/vat.js:35]: To be clear, it was already silenced (by never being type-checked) so this is not a regression. The goal of this PR is to `checkJs` and unblock other `checkJs` PRs, not to solve every type error.

I leave it up to the package maintainers to prioritize fixing this. If you want I can change the comment to include FIXME or a link to a ticket someone requests that I include here.

File: packages/SwingSet/src/kernel/slogger.js
Comment (turadg) [reply packages/SwingSet/src/kernel/slogger.js:285]: Yeah. If there's a ticket to do that I'll include it in the comment

File: packages/SwingSet/src/lib-nodejs/worker-protocol.js
Comment (turadg) [reply packages/SwingSet/src/lib-nodejs/worker-protocol.js]: Agreed.

File: packages/SwingSet/src/liveslots/liveslots.js
Comment (turadg) [reply packages/SwingSet/src/liveslots/liveslots.js:1499]: To unblock the other `checkJs` PRs I propose that this land with the FIXME comment and the package maintainers prioritize fixing it as they wish.

File: packages/SwingSet/src/vats/comms/state.js
Comment (turadg) [reply packages/SwingSet/src/vats/comms/state.js]: Thanks for explaining that. A quick fix was to update the `insistCapData` assertion.

**Review by mhofman — APPROVED** [review]
Thanks for addressing that feedback!

**Pull request merged**
